Adding discarded TPs metric

xinyue-uoft opened the pull request:
This pull request introduces new metric to track TPs that are discarded by the `TPStreamWriterModule`. This change adds two new fields to `opmon::TPStreamWriter`.

### Discarded Tardy TPs

The current counter introduced keeps track of the incremental and total number of TPs discarded when exception arise and 'IgnorableDataStoreProblem' is true. In other words, only those associated with TardyTPsDiscarded warning is counted. In terms of code organization, the additional metric is analogous to the already present tps received and written.

### Tests
- passes 'dbt-unittest-summary'
- passes 'integtest/minimal_system_quick_test'

Commit:
Add new types to TPStreamWriterInfo.

Commit:
Add tps_discarded counters, only collect triggers discarded due to IgnorableDataStoreProblem.

mroda88 reviewed (changes requested):
The total counter should be reset to 0 in `do_start`.

Commit:
Initialize m_total_tps_discard in do_start for TPStreamWriterModule.

xinyue-uoft commented:
Tests

1. passes 'dbt-unittest-summary'
2. passes 'integtest/minimal_system_quick_test'

Commit:
Remove extra number of TPs calculation. Fix behaviour to always have discarded TPs metric independent of m_warn_user_when_tardy_tps_are_discarded.

Commit:
Moving declaration of number_of_tps out of the try block, so that catch block can access its value

xinyue-uoft commented:
Renamed `number_of_tps_written` to `number_of_tps`, this is because it can be used again in handling discarded tps, therefore no need to do the same calculation at two different placed. Its declaration is moved out of the try block inside the do while loop. This is due to the try block and catch block not in the same scope, so that local variable declared and set in try can not be accessed in catch.

Fixed the behaviour that number of discarded TPs is not recorded when `m_warn_user_when_tardy_tps_are_discarded` is `false`. Whether we throw a warning does not change the fact that TP was discarded.

Tests

1. passes 'dbt-unittest-summary'
2. passes 'integtest/minimal_system_quick_test'

Commit:
Add retry_wait_usec logic change by Michal

xinyue-uoft commented:
Tests

1. passes 'dbt-unittest-summary'
2. passes 'integtest/minimal_system_quick_test'

mroda88 commented on plugins/TPStreamWriterModule.cpp:
```diff
@@ -245,29 +248,29 @@ TPStreamWriterModule::do_work(std::atomic<bool>& running_flag)
         should_retry = false;
+	size_t number_of_tps = 0;
         try {
+	  number_of_tps = (timeslice_ptr->get_sum_of_fragment_payload_sizes() / sizeof(trgdataformats::TriggerPrimitive));
```
this outside the try block.

mroda88 commented on plugins/TPStreamWriterModule.cpp:
```diff
@@ -245,29 +248,29 @@ TPStreamWriterModule::do_work(std::atomic<bool>& running_flag)
           double seconds_too_late = m_accumulation_interval_seconds * timeslice_number_diff;
           m_tardy_timeslice_max_seconds = std::max(m_tardy_timeslice_max_seconds.load(), seconds_too_late);
+	  m_tps_discarded += number_of_tps;
+	  m_total_tps_discarded += number_of_tps;
```
We need the increments also in the DataWriting problem catch, a few lines down the code.

Commit:
Move setting of number_of_tps to outside of try block.

Commit:
Increment of counter in case of DataWritingProblem.

xinyue-uoft commented:
Current behaviour is to increment tps discarded and total tps discarded counters when all of these are true: 1. Exception occurs. 2. The exception is not a `RetryableDataStoreProblem`.

Tests:

1. passes 'dbt-unittest-summary'
2. passes 'integtest/minimal_system_quick_test'

MRiganSUSX reviewed (approved):
LGTM.

The code is correct.
Passes unit-tests.
Passes minimal_system_quick_test.

@mroda88 has also implemented a plot for these metrics as part of the `Trigger Primitives` dashboard, and we see the metrics to be published/updated.

Commit:
Update version to 4.1.3

xinyue-uoft commented:
Updated dfmodules VERSION 4.1.2 to dfmodules VERSION 4.1.3.
